src/quicklist.c

```diff
@@ -698,9 +788,29 @@ void quicklistReplaceEntry(quicklist *quicklist, quicklistEntry *entry,
  * Returns 1 if replace happened.
  * Returns 0 if replace failed and no changes happened. */
 int quicklistReplaceAtIndex(quicklist *quicklist, long index, void *data,
-                            int sz) {
+                            size_t sz) {
     quicklistEntry entry;
     if (likely(quicklistIndex(quicklist, index, &entry))) {
+
+        int isCurrentPlain = (QL_NODE_IS_PLAIN(entry.node));
+        int isNewPlain = isLargeElement(sz);
+
+        if (isNewPlain && isCurrentPlain) {
+            zfree(entry.node->entry);
+            entry.node->entry = zmalloc(sz);
+            memcpy(entry.node->entry, data, sz);
+            entry.node->sz = sz;
+            quicklistCompress(quicklist, entry.node);
+            return 1;
+        }
+
+        if (isNewPlain || isCurrentPlain) {
+            quicklistInsertAfter(quicklist, &entry, data, sz);
+            quicklistDelRange(quicklist, index, 1);
```

Comment:
We can use `quicklistDelIndex` instead of `quicklistDelRange` to avoid excess traversal.